### PR TITLE
refactor(router): added a new type minor unit to amount

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Currently we have amount as i64 value, everywhere but we want it to be inside some Unit struct stating its denomination type. With this change we are adding MinorUnit as a new type which will flow through all core flows, and will send this type to connectors in upcoming pr's

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Flows to be tested

- normal payments create -> payments confirm (all prod connectors)
- capture flow
- surcharge flow
- mandate flow
- payouts
- FRM flows

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -689,28 +691,80 @@ impl PaymentsRequest {
     }
 }
 
+#[derive(Default, Debug, serde::Deserialize, serde::Serialize, Clone, Copy, PartialEq, Eq)]
+pub struct MinorUnit(i64);
+
+impl MinorUnit {
+    pub fn get_amount_as_i64(&self) -> i64 {
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:698]: Don't expose this function

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -689,28 +691,80 @@ impl PaymentsRequest {
...
+
+    pub fn add(&self, a2: Self) -> Self {
+        Self::new(self.get_amount_as_i64() + a2.get_amount_as_i64())
+    }
+
+    pub fn substract(&self, a2: Self) -> Self {
+        Self::new(self.get_amount_as_i64() - a2.get_amount_as_i64())
+    }
+
+    pub fn get_optional_amount_as_i64(optional_amount: Option<Self>) -> Option<i64> {
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:710]: This function also should be removed

**Review by jarnura — DISMISSED** [review]

**Review by vspecky — DISMISSED** [review]
on behalf of routing, LGTM

**Review by SamraatBansal — DISMISSED** [review]
LGTM from connectors code changes

**Review by apoorvdixit88 — APPROVED** [review]
Dashboard specific changes are fine!

File: crates/router/src/connector/utils.rs [hunk 3]
```diff
@@ -145,7 +151,9 @@ where
                 let captured_amount =
                     types::Capturable::get_captured_amount(&self.request, payment_data);
                 let total_capturable_amount = payment_data.payment_attempt.get_total_amount();
-                if Some(total_capturable_amount) == captured_amount {
+                if total_capturable_amount
```
Comment (hrithikesh026) [inline crates/router/src/connector/utils.rs:154]: This condition will fail is total_capturable_amount is 0 and captured_amount is None.

File: crates/router/src/types/storage/payment_attempt.rs [hunk 4]
```diff
@@ -66,8 +67,12 @@ impl PaymentAttemptExt for PaymentAttempt {
...
         })
     }
-    fn get_total_amount(&self) -> i64 {
-        self.amount + self.surcharge_amount.unwrap_or(0) + self.tax_amount.unwrap_or(0)
+    fn get_total_amount(&self) -> MinorUnit {
+        self.amount.add(
+            self.surcharge_amount
+                .unwrap_or_default()
+                .add(self.tax_amount.unwrap_or_default()),
+        )
```
Comment (hrithikesh026) [inline crates/router/src/types/storage/payment_attempt.rs:75]:
```suggestion
        self.amount
            .add(self.surcharge_amount.unwrap_or_default())
            .add(self.tax_amount.unwrap_or_default()),
        
```

File: crates/router/src/connector/utils.rs (hunk 3 above)
Comment (sahkal) [reply crates/router/src/connector/utils.rs:154]: if captured amount it will default to 0 right, so the comparison will be between 

let's say total capturable amount is 100

`MinorUnit(100) == MinorUnit(0)` which will move it to else condition

**Review by hrithikesh026 — APPROVED** [review]

**Review by vspecky — DISMISSED** [review]
LGTM on behalf of routing

**Review by apoorvdixit88 — DISMISSED** [review]

**Review by SamraatBansal — DISMISSED** [review]

File: crates/router/src/services/kafka/payment_attempt.rs [hunk 5]
```diff
@@ -60,14 +61,14 @@ impl<'a> KafkaPaymentAttempt<'a> {
             merchant_id: &attempt.merchant_id,
             attempt_id: &attempt.attempt_id,
             status: attempt.status,
-            amount: attempt.amount,
+            amount: attempt.amount.get_amount_as_i64(),
```
Comment (Narayanbhat166) [inline crates/router/src/services/kafka/payment_attempt.rs:64]: This should be minorUnit

File: crates/router/src/core/payments/types.rs [hunk 6]
```diff
@@ -182,33 +185,40 @@ impl MultipleCaptureData {
...
         let surcharge_amount = request_surcharge_details.surcharge_amount;
-        let tax_on_surcharge_amount = request_surcharge_details.tax_amount.unwrap_or(0);
+        let tax_on_surcharge_amount = request_surcharge_details.tax_amount.unwrap_or_default();
         Self {
             original_amount: payment_attempt.amount,
-            surcharge: common_types::Surcharge::Fixed(request_surcharge_details.surcharge_amount),
+            surcharge: common_types::Surcharge::Fixed(
+                request_surcharge_details
+                    .surcharge_amount
+                    .get_amount_as_i64(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/types.rs:213]: can you check this?

File: crates/router/src/core/payouts.rs [hunk 7]
```diff
@@ -368,8 +368,10 @@ pub async fn payouts_update_core(
 
     // Update DB with new data
     let payouts = payout_data.payouts.to_owned();
+    let amount = common_utils::types::MinorUnit::from(req.amount.unwrap_or(api::Amount::Zero))
```
Comment (Narayanbhat166) [inline crates/router/src/core/payouts.rs:371]: we should consider payout amount also here

File: crates/router/src/core/payments/operations/payments_incremental_authorization.rs [hunk 8]
```diff
@@ -75,7 +75,7 @@ impl<F: Send + Clone>
             })?
         }
 
-        if request.amount < payment_intent.amount {
+        if request.amount.is_greater_than(payment_intent.amount) {
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payments_incremental_authorization.rs:78]: this should be is_less_than

File: crates/common_utils/src/types.rs [hunk 9]
```diff
@@ -212,3 +225,92 @@ where
...
+    fn sub(self, a2: Self) -> Self {
+        Self(self.0 - a2.0)
+    }
+}
+
+impl PartialOrd for MinorUnit {
+    fn partial_cmp(&self, a2: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.0.cmp(&a2.0))
+    }
+}
```
Comment (Narayanbhat166) [inline crates/common_utils/src/types.rs:316]: you can derive PartialOrd

File: crates/router/src/types/transformers.rs [hunk 10]
```diff
@@ -938,7 +939,7 @@ impl ForeignFrom<storage::Capture> for payments::CaptureResponse {
         Self {
             capture_id: capture.capture_id,
             status: capture.status,
-            amount: capture.amount,
+            amount: MinorUnit::new(capture.amount),
```
Comment (Narayanbhat166) [inline crates/router/src/types/transformers.rs:942]: have we not changed in capture storage?

**Review by jarnura — DISMISSED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by vspecky — APPROVED** [review]
Approved on behalf of @juspay/hyperswitch-routing

File: crates/router/src/core/payments/operations/payments_incremental_authorization.rs [hunk 11]
```diff
@@ -75,7 +75,7 @@ impl<F: Send + Clone>
             })?
         }
 
-        if request.amount < payment_intent.amount {
```
Comment (hrithikesh026) [inline crates/router/src/core/payments/operations/payments_incremental_authorization.rs:78]: Why was this needed?

**Review by SamraatBansal — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]
